### ISSUE TITLE
PersistenceDelegator assumes wrong column name on join table

### ISSUE THREAD

**Issue by xshaka** [Impetus/kundera#791]
I have found a weird behavior when having join tables in many to many relations, and turns out that the cause is this one:

```
private void flushJoinTableData()
{
    if (applyFlush())
    {
        for (JoinTableData jtData : flushManager.getJoinTableData())
        {
            if (!jtData.isProcessed())
            {
                EntityMetadata m = KunderaMetadataManager.getEntityMetadata(kunderaMetadata,
                        jtData.getEntityClass());
                Client client = getClient(m);
                if (OPERATION.INSERT.equals(jtData.getOperation()))
                {
                    client.persistJoinTable(jtData);
                    jtData.setProcessed(true);
                }
                else if (OPERATION.DELETE.equals(jtData.getOperation()))
                {
                    for (Object pk : jtData.getJoinTableRecords().keySet())
                    {
                        client.deleteByColumn(m.getSchema(), jtData.getJoinTableName(),
                                ((AbstractAttribute) m.getIdAttribute()).getJPAColumnName(), pk);
                    }
                    jtData.setProcessed(true);
                }
            }

        }
    }
}
```

notice the line:

```
client.deleteByColumn(m.getSchema(), jtData.getJoinTableName(), ((AbstractAttribute) m.getIdAttribute()).getJPAColumnName(), pk);
```

which is saying that the join column name will be the same as the id column name of the class having the relation. Why not to do something like:

```
client.deleteByColumn(m.getSchema(), jtData.getJoinTableName(), jtData.getJoinColumnName(), pk);
```

?

**Comment by karthikprasad13** [comment]
@xshaka 

We are looking into it and will update soon.

-Karthik

**Comment by karthikprasad13** [comment]
@xshaka 

Can you explain the " weird behavior"? It would help us debug the code accordingly.

-Karthik

**Comment by xshaka** [comment]
Ok I'm using Kundera 3.1, take a look at this  entity class:

```
@Entity
@Table(name = "college")
public class College {

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private UUID id;

    @Column(name = "collegeName")
    private String name;

    @OneToMany(cascade = { CascadeType.ALL }, fetch = FetchType.LAZY)
    @JoinColumn(name = "collegeId")
    private Set<Professor> professors = new HashSet<>();

    @OneToMany(cascade = { CascadeType.ALL }, fetch = FetchType.LAZY)
    @JoinColumn(name = "collegeId")
    private Set<Student> students = new HashSet<>();

    @ManyToMany(cascade = {CascadeType.PERSIST, CascadeType.MERGE, CascadeType.REMOVE})
    @JoinTable(name = "college_subject_mapping", joinColumns = @JoinColumn(name = "collegeId"), inverseJoinColumns = @JoinColumn(name = "subjectId"))
    private Set<Subject> taughtSubjects = new HashSet<>();

    public UUID getId() {
        return id;
    }

    public void setId(UUID id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Set<Professor> getProfessors() {
        return professors;
    }

    public void setProfessors(Set<Professor> professors) {
        this.professors = professors;
    }

    public Set<Student> getStudents() {
        return students;
    }

    public void setStudents(Set<Student> students) {
        this.students = students;
    }

    public Set<Subject> getTaughtSubjects() {
        return taughtSubjects;
    }

    public void setTaughtSubjects(Set<Subject> taughtSubjects) {
        this.taughtSubjects = taughtSubjects;
    }
}
```

focus on 'taughtSubjects' ManyToMany mapping. When Kundera tries to flush the join table in order to delete for example, I get this:

```
SELECT "key" FROM "college_subject_mapping" WHERE "id" = a8999630-6e8d-11e5-b33e-cfd49dcaba3c
2015-10-09 10:57:20,439 [main] ERROR com.impetus.kundera.client.cassandra.dsdriver.DSClient - Error while executing query SELECT "key" FROM "college_subject_mapping" WHERE "id" = a8999630-6e8d-11e5-b33e-cfd49dcaba3c.

com.impetus.kundera.KunderaException: com.datastax.driver.core.exceptions.InvalidQueryException: Undefined name id in where clause ('id = a8999630-6e8d-11e5-b33e-cfd49dcaba3c')
    at com.impetus.kundera.client.cassandra.dsdriver.DSClient.execute(DSClient.java:724)
    at com.impetus.kundera.client.cassandra.dsdriver.DSClient.getColumnsById(DSClient.java:361)
    at com.impetus.kundera.client.cassandra.dsdriver.DSClient.deleteByColumn(DSClient.java:416)
    at com.impetus.kundera.persistence.PersistenceDelegator.flushJoinTableData(PersistenceDelegator.java:862)
    at com.impetus.kundera.persistence.PersistenceDelegator.execute(PersistenceDelegator.java:833)
    at com.impetus.kundera.persistence.PersistenceDelegator.commit(PersistenceDelegator.java:732)
    at com.impetus.kundera.persistence.KunderaEntityTransaction.commit(KunderaEntityTransaction.java:79)
    at org.springframework.orm.jpa.JpaTransactionManager.doCommit(JpaTransactionManager.java:517)
    at org.springframework.transaction.support.AbstractPlatformTransactionManager.processCommit(AbstractPlatformTransactionManager.java:761)
    at org.springframework.transaction.support.AbstractPlatformTransactionManager.commit(AbstractPlatformTransactionManager.java:730)
    at org.springframework.transaction.interceptor.TransactionAspectSupport.commitTransactionAfterReturning(TransactionAspectSupport.java:485)
    at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:291)
    at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:96)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
    at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:653)
    at com.shaka.dataaccess.repositories.CollegeRepository$$EnhancerBySpringCGLIB$$79bbca89.deleteEntity(<generated>)
    at com.shaka.DemoPortalApplicationTests.subjectCollectionAssociation(DemoPortalApplicationTests.java:102)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
    at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
    at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
    at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
    at org.springframework.test.context.junit4.statements.RunBeforeTestMethodCallbacks.evaluate(RunBeforeTestMethodCallbacks.java:74)
    at org.springframework.test.context.junit4.statements.RunAfterTestMethodCallbacks.evaluate(RunAfterTestMethodCallbacks.java:85)
    at org.springframework.test.context.junit4.statements.SpringRepeat.evaluate(SpringRepeat.java:86)
    at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
    at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:241)
    at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:87)
    at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
    at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
    at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
    at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
    at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
    at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
    at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61)
    at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:70)
    at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
    at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:180)
    at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
    at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:78)
    at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:212)
    at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:68)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at com.intellij.rt.execution.application.AppMain.main(AppMain.java:140)
Caused by: com.datastax.driver.core.exceptions.InvalidQueryException: Undefined name id in where clause ('id = a8999630-6e8d-11e5-b33e-cfd49dcaba3c')
    at com.datastax.driver.core.exceptions.InvalidQueryException.copy(InvalidQueryException.java:35)
    at com.datastax.driver.core.DefaultResultSetFuture.extractCauseFromExecutionException(DefaultResultSetFuture.java:269)
    at com.datastax.driver.core.DefaultResultSetFuture.getUninterruptibly(DefaultResultSetFuture.java:183)
    at com.datastax.driver.core.AbstractSession.execute(AbstractSession.java:52)
    at com.impetus.kundera.client.cassandra.dsdriver.DSClient.execute(DSClient.java:719)
    ... 49 more
Caused by: com.datastax.driver.core.exceptions.InvalidQueryException: Undefined name id in where clause ('id = a8999630-6e8d-11e5-b33e-cfd49dcaba3c')
    at com.datastax.driver.core.Responses$Error.asException(Responses.java:102)
    at com.datastax.driver.core.DefaultResultSetFuture.onSet(DefaultResultSetFuture.java:118)
    at com.datastax.driver.core.RequestHandler.setFinalResult(RequestHandler.java:183)
    at com.datastax.driver.core.RequestHandler.access$2300(RequestHandler.java:45)
    at com.datastax.driver.core.RequestHandler$SpeculativeExecution.setFinalResult(RequestHandler.java:748)
    at com.datastax.driver.core.RequestHandler$SpeculativeExecution.onSet(RequestHandler.java:573)
    at com.datastax.driver.core.Connection$Dispatcher.channelRead0(Connection.java:991)
    at com.datastax.driver.core.Connection$Dispatcher.channelRead0(Connection.java:913)
    at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105)
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:333)
    at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:319)
    at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:254)
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:333)
    at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:319)
    at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:333)
    at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:319)
    at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:163)
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:333)
    at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:319)
    at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:787)
    at io.netty.channel.epoll.EpollSocketChannel$EpollSocketUnsafe.epollInReady(EpollSocketChannel.java:722)
    at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:326)
    at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:264)
    at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:116)
    at java.lang.Thread.run(Thread.java:745)
```

Now, see the CQL statement Kundera is trying to execute:

```
SELECT "key" FROM "college_subject_mapping" WHERE "id" = a8999630-6e8d-11e5-b33e-cfd49dcaba3c
```

Clearly, 'id' column doesn't exist in my join table, and when digging into the code, I've found what I've described above: PersistentDelegator wrongly assumes that join column name to be the same as the id column of the class (bug is to use this **((AbstractAttribute) m.getIdAttribute()).getJPAColumnName()**):

```
client.deleteByColumn(m.getSchema(), jtData.getJoinTableName(), ((AbstractAttribute) m.getIdAttribute()).getJPAColumnName(), pk);
```

**Comment by chhavigangwal** [comment]
@xshaka 

Seems to be some issue with CQL2 & CQL3 here, will update you ASAP

Chhavi

**Comment by karthikprasad13** [comment]
@xshaka 

You don't need to initialize the fields in the entity, i.e:

```
private Set<Subject> taughtSubjects = new HashSet<>();
```

can be changed to:

```
 private Set<Subject> taughtSubjects;
```

Also, I verified CRUD of objects that you shared and seems to work fine. I am trying to reproduce the issue at my end, can you share when exactly you got the issue? Sample query or operation would be great.

-Karthik.

**Comment by xshaka** [comment]
Try messing around with the collection having the @JoinTable annotation and make sure the join columns have a name different to their referenced @Id columns.

**Comment by karthikprasad13** [comment]
@xshaka 

Cannot reproduce the issue. Can you share a sample project?

-Karthik

**Comment by chhavigangwal** [comment]
Fixed with Kundera-3.3
